qa/tasks/cephfs/test_admin.py

```diff
@@ -16,15 +16,72 @@
                                       gen_osd_cap_str, gen_mds_cap_str)
 
 log = logging.getLogger(__name__)
+MDS_RESTART_GRACE = 60
 
 class TestLabeledPerfCounters(CephFSTestCase):
     CLIENTS_REQUIRED = 2
     MDSS_REQUIRED = 1
 
-    def test_per_client_labeled_perf_counters(self):
+    def test_per_client_labeled_perf_counters_on_client_connect(self):
         """
-        That the per-client labelled perf counters depict the clients
-        performaing IO.
+        That the per-client labelled metrics are generated during client connect
+        """
+        self.mount_a.teardown()
+        self.mount_a.mount_wait(cephfs_name=self.fs.name)
+        mount_a_id = f'client.{self.mount_a.get_global_id()}'
+        with safe_while(sleep=1, tries=30, action=f'wait for counters - {mount_a_id}') as proceed:
+            while proceed():
+                dump = self.fs.rank_tell(["counter", "dump"])
+                per_client_metrics_key = f"mds_client_metrics-{dump['mds_client_metrics'][0]['labels']['fs_name']}"
+                clients = [c["labels"]["client"] for c in dump.get(per_client_metrics_key, {})]
+                if mount_a_id in clients:
+                    # success, got metrics.
+                    break
+
+    def test_per_client_labeled_perf_counters_on_client_disconnect(self):
+        """
+        That the per-client labelled metrics are unavailable during client disconnect
+        """
+        mount_a_id = f'client.{self.mount_a.get_global_id()}'
+        self.mount_a.teardown()
+        with safe_while(sleep=1, tries=30, action=f'wait for counters - {mount_a_id}') as proceed:
+            while proceed():
+                dump = self.fs.rank_tell(["counter", "dump"])
+                per_client_metrics_key = f"mds_client_metrics-{dump['mds_client_metrics'][0]['labels']['fs_name']}"
+                clients = [c["labels"]["client"] for c in dump.get(per_client_metrics_key, {})]
+                if clients and mount_a_id not in clients:
+                    # success, no metrics.
+                    break
+
+    def test_per_client_labeled_perf_counters_on_client_reconnect(self):
+        """
+        That the per-client labelled metrics are generated during client reconnect
+        """
+        fs_name = self.fs.name
+        self.fs.fail()
+        self.mount_a.kill()
+        self.mount_a.kill_cleanup()
+        self.mount_b.kill()
+        self.mount_b.kill_cleanup()
+        self.fs.set_joinable()
+        log.info("Waiting for reconnect...")
+        self.fs.wait_for_state("up:reconnect", timeout=MDS_RESTART_GRACE)
+        log.info("Waiting for active...")
+        self.fs.wait_for_state('up:active', timeout=MDS_RESTART_GRACE)
+        self.mount_a.mount_wait(cephfs_name=f'{fs_name}')
```

Comment:
Are you sure this is reconnecting? kill_cleanup does a unmount_wait(force=True) so the mounts are unmounted. For checking reconnections, you'd need to disrupt the network b/w the client and the MDS (something like failing over the MDS would trigger a reconnect from the client).

Reply:
Also, post reconnecting, validate the IO perf counters when the client does IO.

Reply:
Fixed